techlibs/quicklogic/synth_quicklogic.cc

```diff
@@ -108,51 +109,94 @@ struct SynthQuickLogicPass : public ScriptPass
             run("peepopt");
             run("pmuxtree");
-            run("opt");
-            run("memory -nomap");
             run("opt_clean");
+            run("share");
+            run("techmap -map +/cmp2lut.v -D LUT_WIDTH=4");
+            run("opt_expr");
+            run("opt_clean");
+
             run("alumacc");
+            run("opt");
+            run("fsm");
             run("opt -fast");
+            run("memory -nomap");
             run("opt_clean");
-            run("opt -full");
-            run("check");
         }
 
-        if (check_label("map")) {
-            if (check_label("map_bram", "(skip if -nobram)"))
-            {
-               run("memory_bram -rules +/quicklogic/" + family + "_brams.txt");
-               run("techmap -map +/quicklogic/" + family + "_brams_map.v");
-            }
+        if (check_label("map_bram", "(skip if -nobram)")) {
+            run("memory_bram -rules +/quicklogic/" + family + "_brams.txt");
+            run("techmap -map +/quicklogic/" + family + "_brams_map.v");
+        }
+		
+        if (check_label("map_ffram")) {
+            run("opt -fast -mux_undef -undriven -fine");
+            run("memory_map -iattr -attr !ram_block -attr !rom_block -attr logic_block "
+                    "-attr syn_ramstyle=auto -attr syn_ramstyle=registers "
+                    "-attr syn_romstyle=auto -attr syn_romstyle=logic");
+            run("opt -undriven -fine");
+        }
+
+        if (check_label("map_gates")) {
             run("techmap");
             run("opt -fast");
-            if (family == "pp3") {
-                run("muxcover -mux8 -mux4");
-            }
-
+            run("muxcover -mux8 -mux4");
             run("opt_expr -clkinv");
-            run("dff2dffe");
             run("opt -fast");
-            run("techmap -map +/quicklogic/ffs_map.v");
+            run("opt_expr");
+            run("opt_merge");
+            run("opt_rmdff");
+            run("opt_clean");
+            run("opt");
+        }
 
-            if (family == "pp3") {
-                run("abc -luts 1,2,2");
-            }
-            run("opt -fast");
+        if (check_label("map_ffs")) {
+			run("opt_expr -clkinv");
+            run("dff2dffe");
 
-            run("techmap -map +/quicklogic/lut_map.v");
+            std::string techMapArgs = " -map +/quicklogic/" + family + "_ffs_map.v";
+            run("techmap " + techMapArgs);
+            run("opt_expr -mux_undef");
+            run("simplemap");
+            run("opt_expr");
+            run("opt_merge");
+            run("opt_rmdff");
             run("opt_clean");
+            run("opt");
+        }
 
-            run("techmap -map +/quicklogic/cells_map.v");
-            run("opt_clean");
+        if (check_label("map_luts")) {
+            std::string techMapArgs = " -map +/quicklogic/" + family + "_latches_map.v";
+            run("techmap " + techMapArgs);
+            run("abc -lut 4"); // -luts 1,2,2
```

Comment:
This should be left as it was (`abc -luts 1,2,2`). Our experiments have shown that allowing ABC to infer LUT4 yields in much bigger designs.

Reply:
I asked this question last week and was not sure why. Ok, I will revert this change.